### DISCUSSION TITLE
cache is locked, waiting for it to be released

### DISCUSSION THREAD

**Discussion by tacman** [symfony/symfony#52934] [Q&A]
Is there a way to forcefully unlock the cache?

I'm stumped as to why the cache keeps locking -- the code seems pretty straightforward, a set of http fetches.  Since they're slow, I cache the response.

I'm not doing anything tricky, running them in parallel or a separate process, just iterating through a list. 

And it'll be working fine for a while, then it freezes up.  I've tried different caches (redis, apcu, files), I've cleared the cache and the cache:pool, run the redis-cli FLUSHALL command, then it works for a while and then locks.

I'm using Symfony 6.4, PHP 8.3, and nothing seems too out of the ordinary.  I'm not even sure where to start looking on how to fix this.

Thx.

```bash
  "key" => "detailsB25-0033"
]
19:57:13 INFO      [http_client] Request: "GET https://lims.dccouncil.gov/api/v2/PublicData//LegislationDetails/B25-0033"
19:57:14 INFO      [http_client] Response: "200 https://lims.dccouncil.gov/api/v2/PublicData//LegislationDetails/B25-0033"
19:57:17 INFO      [cache] Lock acquired, now computing item "detailsB25-0034"
[
  "key" => "detailsB25-0034"
]
19:57:17 INFO      [http_client] Request: "GET https://lims.dccouncil.gov/api/v2/PublicData//LegislationDetails/B25-0034"
19:57:17 INFO      [http_client] Response: "200 https://lims.dccouncil.gov/api/v2/PublicData//LegislationDetails/B25-0034"
19:57:20 INFO      [cache] Lock acquired, now computing item "detailsB25-0035"
[
  "key" => "detailsB25-0035"
]
19:57:20 INFO      [http_client] Request: "GET https://lims.dccouncil.gov/api/v2/PublicData//LegislationDetails/B25-0035"
19:57:20 INFO      [http_client] Response: "200 https://lims.dccouncil.gov/api/v2/PublicData//LegislationDetails/B25-0035"
19:57:23 INFO      [cache] Lock acquired, now computing item "detailsB25-0036"
[
  "key" => "detailsB25-0036"
]
19:57:23 INFO      [http_client] Request: "GET https://lims.dccouncil.gov/api/v2/PublicData//LegislationDetails/B25-0036"
19:57:23 INFO      [http_client] Response: "200 https://lims.dccouncil.gov/api/v2/PublicData//LegislationDetails/B25-0036"
19:57:26 INFO      [cache] Item "detailsB25-0037" is locked, waiting for it to be released
[
  "key" => "detailsB25-0037"
```

**Comment by MatTheCat** [comment]
It may boil down to #27604:

> For each apps, there can be at most 20 concurrent processes that compute items at the same time and only one per cache-key.

Have a look at #28588

**Reply by tacman** [reply]
So that sounds right, that it goes about 20 calls before locking, but then it's completely locked.

BUT the calculated cache isn't running concurrently.  It's a single process, a loop over a set of URL's.  I've used the cache in a bunch of apps and haven't had this issue, but this app used to be in 6.1 and was upgraded to 6.4.

I should learn how to use the [HttpCachingClient](https://jolicode.com/blog/aggressive-caching-with-symfony-http-client), since that's really the behavior I want, fetch a bunch of JSON urls, but many of them can have long cache times.  

This client may someday be easier to configure: https://github.com/symfony/symfony/issues/41843

**Reply by MatTheCat** [reply]
Hm so this seems like locks are not released between requests :thinking: 

What Symfony version are you running? What SAPI are you using?

**Comment by tacman** [comment]
and it appears to not respect the timeout when it's locked! :-(

**Comment by tacman** [comment]
This continues to be a problem for me, it doesn't matter which cache adapter I use (redis, apcu, filesystem).

I can get further by limiting the number of messages processed, but it still eventually hangs

```bash
while true; do symfony console mess:stats && symfony console messenger:consume async -vv --limit 8 --time-limit 90; sleep 2;   done;
                                                      
 [OK] Consuming messages from transport "async".                                                                        
                                                                                                                       

 // The worker will automatically exit once it has processed 8 messages, been running for 90s or received a stop signal
 // via the messenger:stop-workers command.                                                                            

 // Quit the worker with CONTROL-C.                                                                                    

16:14:49 INFO      [cache] Lock acquired, now computing item "zenstruck_messenger_monitor.worker.291970" ["key" => "zenstruck_messenger_monitor.worker.291970"]
16:14:49 INFO      [messenger] Received message App\Message\FetchWikidataMessage ["class" => "App\Message\FetchWikidataMessage"]
16:14:49 INFO      [cache] Lock acquired, now computing item "Q465767en" ["key" => "Q465767en"]
16:14:50 INFO      [app] image Q465767/7bbb7f841abbe7bce472db1e91e1f9fe.jpg already exists in filesystem
16:14:50 INFO      [messenger] Message App\Message\FetchWikidataMessage handled by App\MessageHandler\FetchWikidataMessageHandler::__invoke ["class" => "App\Message\FetchWikidataMessage","handler" => "App\MessageHandler\FetchWikidataMessageHandler::__invoke"]
16:14:50 INFO      [messenger] App\Message\FetchWikidataMessage was handled successfully (acknowledging to transport). ["class" => "App\Message\FetchWikidataMessage"]
16:14:50 INFO      [cache] Lock acquired, now computing item "zenstruck_messenger_monitor.worker.291970" ["key" => "zenstruck_messenger_monitor.worker.291970"]
16:14:50 INFO      [messenger] Received message App\Message\FetchWikidataMessage ["class" => "App\Message\FetchWikidataMessage"]
16:14:50 INFO      [cache] Item "Q547153en" is locked, waiting for it to be released ["key" => "Q547153en"]
```


I'm running php 8.3, the latest Symfony CLI, and it doesn't matter if I use bin/console or symfony console, or if I disable the zenstruck_messenger_monitor.

**Reply by MatTheCat** [reply]
Does [`LockRegistry::compute`’s `finally` block](https://github.com/symfony/symfony/blob/5eb2e615417b0a28d123ed28bc8e6ae3a311638a/src/Symfony/Component/Cache/LockRegistry.php#L127-L130) ever run in your case?

Could you provide a reproducer?

**Reply by tacman** [reply]
where would that happen?  I'm just the cache at a pretty basic level:

```php
    public function fetchWikidataPage(?string $code, string $lang = 'en'): ?Entity
    {
        if (is_null($code)) {
            return null;
        }

        $key = $code . $lang;
        $value = $this->cache->get($key, function (ItemInterface $item) use ($code, $lang) {
            try {
                $content = $this->wikidata->get($code, $lang); // makes the API call to wikidata.org
            } catch (\Exception $exception) {
                // @todo: log error
                return null;
            }
            return $content;
        });
        return $value;
    }

```

Working on a reproducer.  The issue _seems_ to be when I run the script that loads the data and then interrupt it, but not sure.

**Reply by MatTheCat** [reply]
> where would that happen?

It should happen when a cache item is computed.

>  The issue seems to be when I run the script that loads the data and then interrupt it

Don't know a lot about `flock` but it sounds like it could be an explanation.

**Reply by tacman** [reply]
I don't think the finally block is ever called, as it seems to be hanging there in that loop.

**Reply by MatTheCat** [reply]
Iʼve read that locks are not automatically released on interrupt (https://github.com/phpstan/phpstan/issues/1979#issuecomment-472358700 e.g.) so this seems to be what you're experiencing.

A reproducer would be welcome though!
